### PR TITLE
Ports tgs runechat in replacement to our current overhead chat

### LINKED ISSUE #2243: Runespace Chat subsystem
Something like this but for beestation runescape chat.
https://github.com/tgstation/tgstation/pull/52425

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/92213934-runespace-chat-subsystem?utm_campaign=plugin&utm_content=tracker%2F120077027&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F120077027&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #3764: cant sing on the radio

### PR DESCRIPTION
**Pull request by Archanial**
## About The Pull Request

Ported PRs:
https://github.com/tgstation/tgstation/pull/46499
https://github.com/tgstation/tgstation/pull/53217
https://github.com/tgstation/tgstation/pull/48831
https://github.com/tgstation/tgstation/pull/53570
https://github.com/tgstation/tgstation/pull/53809
https://github.com/tgstation/tgstation/pull/55420
https://github.com/tgstation/tgstation/pull/54522
https://github.com/tgstation/tgstation/pull/54318
https://github.com/tgstation/tgstation/pull/51790
https://github.com/tgstation/tgstation/pull/50608
https://github.com/tgstation/tgstation/pull/50830
https://github.com/tgstation/tgstation/pull/52212
https://github.com/tgstation/tgstation/pull/52425
https://github.com/tgstation/tgstation/pull/52479
https://github.com/tgstation/tgstation/pull/52584
https://github.com/tgstation/tgstation/pull/52738
https://github.com/tgstation/tgstation/pull/52959
https://github.com/tgstation/tgstation/pull/53570
https://github.com/tgstation/tgstation/pull/53809
https://github.com/tgstation/tgstation/pull/54318

closes: #3764 
closes: #2243

- [x] port stuff
- [x] make it work
- [x] port bacons formatting prs
- [x] make filesaves work
- [x] fix bugs
- [x] test
- [x] write changelog, credit original creators, add images to PR.

## Why It's Good For The Game

Some images:
obraz
obraz
obraz
obraz
obraz
obraz
obraz
obraz
obraz
obraz
obraz
obraz

Also an example of maptext:
obraz

Chat is now handled by subsystem. 
Overhead chat is not limited to living.
Overhead images when speaking different languages.
Lots of (probably) good refactors.
Hear now takes `list/message_mods = list())` instead of `message_mode`.
aiEye is now ai_eye

## Changelog 
:cl:bobbahbrown, Rohesie, Mothblocks, Azarak, LemonInTheDark, Wayland-Smithy, Couls, nightred
add: New overhead chat ported from TG.
del: Old overhead chat is gone.
fix: Overhead chat doesn't render under plants.
add: Overhead chat for non /living.
add: New in client preferences regarding visibility of the chat, chat displaying emotes and maximum length of the message.
tweak: More UI elements will now be displayed in a non-blurry font.
/:cl:

### CODE REVIEW

File: code/__HELPERS/_lists.dm [hunk 1]
```diff
@@ -21,34 +21,50 @@
...
-			__BIN_MID = (__BIN_LEFT + __BIN_RIGHT) >> 1;\
+			__BIN_ITEM = COMPTYPE;\
+			__BIN_MID = __BIN_ITEM.##COMPARISON > COMPARE.##COMPARISON ? __BIN_MID : __BIN_MID + 1;\
+			__BIN_LIST.Insert(__BIN_MID, INPUT);\
 		};\
-		__BIN_ITEM = LIST[__BIN_MID];\
-		__BIN_MID = __BIN_ITEM.##COMPARE > IN.##COMPARE ? __BIN_MID : __BIN_MID + 1;\
-		LIST.Insert(__BIN_MID, IN);\
-	}
+	} while(FALSE)
```
Comment (PowerfulBacon) [inline code/__HELPERS/_lists.dm:67]: Is our binary inserts used in stat and stuff updated to account for this?

File: code/game/machinery/computer/camera_advanced.dm [hunk 2]
```diff
@@ -5,7 +5,7 @@
 	icon_keyboard = "security_key"
 	var/list/z_lock = list() // Lock use to these z levels
 	var/lock_override = NONE
-	var/mob/camera/aiEye/remote/eyeobj
+	var/mob/camera/ai_eye/remote/eyeobj
```
Comment (PowerfulBacon) [inline code/game/machinery/computer/camera_advanced.dm:8]: ai_eye change is unrelated to this PR and should not be included with this PR. It will cause annoyances when porting.

File: code/datums/chatmessage.dm [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+  * * name - The name to generate a color for
+  * * sat_shift - A value between 0 and 1 that will be multiplied against the saturation
+  * * lum_shift - A value between 0 and 1 that will be multiplied against the luminescence
+  */
+/datum/chatmessage/proc/colorize_string(name, sat_shift = 1, lum_shift = 1)
+	// seed to help randomness
+	var/static/rseed = rand(1,26)
+
+	// get hsl using the selected 6 characters of the md5 hash
+	var/hash = copytext(md5(name + GLOB.round_id), rseed, rseed + 6)
```
Comment (PowerfulBacon) [inline code/datums/chatmessage.dm:295]: md5 is very slow and the thing that absolutely destroyed the server when the stat panel had it.

**Review by PowerfulBacon — COMMENTED** [review]
icons/UI_Icons/chat/chat_icons.dmi - Empty file

File: code/__HELPERS/_lists.dm (hunk 1 above)
Comment (Archanial) [reply code/__HELPERS/_lists.dm:67]: I might be wrong, but it uses yet another binary insert, that was last modified by you, and I left it untouched.

File: code/game/machinery/computer/camera_advanced.dm (hunk 2 above)
Comment (Archanial) [reply code/game/machinery/computer/camera_advanced.dm:8]: Porting from where? TG? They have this name changed too.

File: code/datums/chatmessage.dm (hunk 3 above)
Comment (Archanial) [reply code/datums/chatmessage.dm:295]: This proc is only called if atom has no predefined color or it has no color and its name has changed.
I don't think it's gonna lag the server, especially if being called that rarely. 
How did you solve that in stat panel?

Comment (PowerfulBacon) [reply code/datums/chatmessage.dm:295]: Just didnt use it, the hash was just the name of the overlays themselves.

File: code/game/machinery/computer/camera_advanced.dm (hunk 2 above)
Comment (PowerfulBacon) [reply code/game/machinery/computer/camera_advanced.dm:8]: well ok then

**Comment by Archanial** [comment]
> icons/UI_Icons/chat/chat_icons.dmi - Empty file

It's not empty, no idea why github says it's empty.

**Comment by MNarath1** [comment]
i actually don't really like how tg's overhead chat handles stuff like emotes no one needs 200 cought and sneezes to pop up in overhead chat

**Comment by MNarath1** [comment]
Does it still show for invisible persons too?

File: tools/Runtime Condenser/Output.txt [hunk 4]
```diff
@@ -349,7 +349,7 @@ proc name: get fuel (/obj/item/weapon/weldingtool/proc/get_fuel)
 
 /obj/machinery/power/grounding_rod - 1 time(s).
 
-/mob/camera/aiEye - 1 time(s).
```
Comment (francinum) [inline tools/Runtime Condenser/Output.txt]: Why the fuck do we still have this file and it's contents.

**Review by francinum — DISMISSED** [review]
Seems to work perfectly well under test.

File: code/controllers/subsystem/runechat.dm [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	/// Queue used for storing messages that are scheduled for deletion too far in the future for the buckets
+	var/list/datum/chatmessage/second_queue = list()
+
+/datum/controller/subsystem/runechat/PreInit()
+	bucket_list.len = BUCKET_LEN
+	head_offset = world.time
+	bucket_resolution = world.tick_lag
+
+/datum/controller/subsystem/runechat/stat_entry(msg)
+	..("ActMsgs:[bucket_count] SecQueue:[length(second_queue)]")
```
Comment (park66665) [inline code/controllers/subsystem/runechat.dm]:
```suggestion
	. = ..("ActMsgs:[bucket_count] SecQueue:[length(second_queue)]")
```

File: code/modules/mob/living/say.dm [hunk 6]
```diff
@@ -95,16 +92,24 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
...
+
+	if(message_mods[RADIO_EXTENSION] == MODE_ADMIN)
+		if(client)
+			client.cmd_admin_say(message)
+		return
+
+	if(message_mods[RADIO_EXTENSION] == MODE_DEADMIN)
+		if(client)
+			client.dsay(message)
+		return
```
Comment (park66665) [inline code/modules/mob/living/say.dm]: These shouldn't be ported per #3686

File: code/__DEFINES/say.dm [hunk 7]
```diff
@@ -21,13 +25,20 @@
...
 #define MODE_DEPARTMENT "department"
 #define MODE_KEY_DEPARTMENT "h"
 #define MODE_TOKEN_DEPARTMENT ":h"
 
+#define MODE_ADMIN "admin"
+#define MODE_KEY_ADMIN "p"
+
+#define MODE_DEADMIN "deadmin"
+#define MODE_KEY_DEADMIN "d"
+
```
Comment (park66665) [inline code/__DEFINES/say.dm]: ditto

File: code/modules/mob/dead/observer/say.dm [hunk 8]
```diff
@@ -7,12 +15,22 @@
...
+	var/list/message_mods = list()
+	message = get_message_mods(message, message_mods)
+	if(client && (message_mods[RADIO_EXTENSION] == MODE_ADMIN || message_mods[RADIO_EXTENSION] == MODE_DEADMIN))
+		message = trim_left(copytext_char(message, length(message_mods[RADIO_KEY]) + 2))
+		if(message_mods[RADIO_EXTENSION] == MODE_ADMIN)
+			client.cmd_admin_say(message)
+		else if(message_mods[RADIO_EXTENSION] == MODE_DEADMIN)
+			client.dsay(message)
+		return
+
```
Comment (park66665) [inline code/modules/mob/dead/observer/say.dm]: ditto.

File: code/modules/mob/living/carbon/human/say.dm [hunk 9]
```diff
@@ -52,25 +52,22 @@
...
+		return ITALICS | REDUCE_RANGE
+	else if(message_mods[RADIO_EXTENSION] == MODE_DEPARTMENT)
+		if(ears)
+			ears.talk_into(src, message, message_mods[RADIO_EXTENSION], spans, language, message_mods)
+		return ITALICS | REDUCE_RANGE
+	else if(GLOB.radiochannels[message_mods[RADIO_EXTENSION]])
 		if(ears)
-			ears.talk_into(src, message, message_mode, spans, language)
+			ears.talk_into(src, message, message_mods[RADIO_EXTENSION], spans, language, message_mods)
 			return ITALICS | REDUCE_RANGE
```
Comment (park66665) [inline code/modules/mob/living/carbon/human/say.dm]: I seriously think we should consolidate these two `if` clauses into one some day.

File: code/modules/mob/living/simple_animal/parrot.dm [hunk 10]
```diff
@@ -154,25 +154,22 @@
...
+		return ITALICS | REDUCE_RANGE
+	else if(message_mods[RADIO_EXTENSION] == MODE_DEPARTMENT)
+		if(ears)
+			ears.talk_into(src, message, message_mods[RADIO_EXTENSION], spans, language, message_mods)
+		return ITALICS | REDUCE_RANGE
+	else if(message_mods[RADIO_EXTENSION] in GLOB.radiochannels)
 		if(ears)
-			ears.talk_into(src, message, message_mode, spans, language)
+			ears.talk_into(src, message, message_mods[RADIO_EXTENSION], spans, language, message_mods)
 			return ITALICS | REDUCE_RANGE
```
Comment (park66665) [inline code/modules/mob/living/simple_animal/parrot.dm]: The same goes for this...

File: code/modules/mob/living/simple_animal/bot/bot.dm [hunk 11]
```diff
@@ -372,22 +372,19 @@
...
-	if(message_mode in GLOB.radiochannels)
-		Radio.talk_into(src, message, message_mode, spans, language)
+	if(message_mods[MODE_HEADSET])
+		Radio.talk_into(src, message, , spans, language, message_mods)
+		return REDUCE_RANGE
+	else if(message_mods[RADIO_EXTENSION] == MODE_DEPARTMENT)
+		Radio.talk_into(src, message, message_mods[RADIO_EXTENSION], spans, language, message_mods)
+		return REDUCE_RANGE
+	else if(message_mods[RADIO_EXTENSION] in GLOB.radiochannels)
+		Radio.talk_into(src, message, message_mods[RADIO_EXTENSION], spans, language, message_mods)
```
Comment (park66665) [inline code/modules/mob/living/simple_animal/bot/bot.dm]: ...and this.

File: code/modules/mob/living/say.dm [hunk 12]
```diff
@@ -366,51 +363,49 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
...
-		if(message_mode == MODE_HEADSET)
-			imp.radio.talk_into(src, message, , spans, language)
+		if(message_mods[MODE_HEADSET])
+			imp.radio.talk_into(src, message, , spans, language, message_mods)
 			return ITALICS | REDUCE_RANGE
-		if(message_mode == MODE_DEPARTMENT || (message_mode in imp.radio.channels))
-			imp.radio.talk_into(src, message, message_mode, spans, language)
+		if(message_mods[RADIO_EXTENSION] == MODE_DEPARTMENT || (message_mods[RADIO_EXTENSION] in imp.radio.channels))
+			imp.radio.talk_into(src, message, message_mods[RADIO_EXTENSION], spans, language, message_mods)
 			return ITALICS | REDUCE_RANGE
```
Comment (park66665) [inline code/modules/mob/living/say.dm:364]: A good example would be this, probably.

File: code/modules/mob/dead/observer/say.dm [hunk 13]
```diff
@@ -1,3 +1,11 @@
+//Modified version of get_message_mods, removes the trimming, the only thing we care about here is admin channels
+/mob/dead/observer/get_message_mods(message, list/mods)
+	var/key = message[1]
+	if((key in GLOB.department_radio_prefixes) && length(message) > length(key) + 1 && !mods[RADIO_EXTENSION])
+		mods[RADIO_KEY] = lowertext(message[1 + length(key)])
+		mods[RADIO_EXTENSION] = GLOB.department_radio_keys[mods[RADIO_KEY]]
+	return message
+
```
Comment (park66665) [inline code/modules/mob/dead/observer/say.dm]: It seems like this part should also be deleted per #3686

**Comment by Vexylius** [comment]
Merge this plz

File: code/modules/client/preferences_savefile.dm [hunk 14]
```diff
@@ -143,87 +148,94 @@ SAVEFILE UPDATING/VERSIONING - 'Simplified', or rather, more coder-friendly ~Car
...
-
-	S["purchased_gear"]					>> purchased_gear
-	S["equipped_gear"]					>> equipped_gear
+	READ_FILE(S["asaycolor"], asaycolor)
+	READ_FILE(S["ooccolor"], ooccolor)
+	READ_FILE(S["lastchangelog"], lastchangelog)
+	READ_FILE(S["UI_style"], UI_style)
+	READ_FILE(S["overhead_chat"], overhead_chat)
+	READ_FILE(S["outline_color"], outline_color)
+	READ_FILE(S["outline_enabled"], outline_enabled)
```
Comment (park66665) [inline code/modules/client/preferences_savefile.dm:169]: Heads up, you should entirely remove now-obsoleted ~~`outline_enabled`~~ EDIT: `overhead_chat` pref from the codebase if you can.

**Review by park66665 — COMMENTED** [review]
I actually didn't review core parts of this PR (namely, refactor of message mode, runechat subsystem and chatmessage datum) since I assume those work well enough.

File: code/datums/chatmessage.dm [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+				var/obj/item/card/id/idcard = H.wear_id
+				var/datum/job/wearer_job = SSjob.GetJob(idcard.GetJobName())
+				if(wearer_job)
+					target.chat_color = wearer_job.chat_color
+				else
+					target.chat_color = GLOB.job_colors_pastel[idcard.GetJobName()]
+				target.chat_color_name = H.name
+			else
+				target.chat_color = COLOR_PERSON_UNKNOWN
+			MultiplyHexColor(target.chat_color, 0.85)
```
Comment (PowerfulBacon) [inline code/datums/chatmessage.dm]: whats the point of this?

File: code/datums/chatmessage.dm [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+		else if(target.chat_color_name != target.name)		//this is only needed if target's name has changed since we last set it's color
+			var/mob/living/L = target
+			target.chat_color = L.chat_color
+			target.chat_color_name = L.name
```
Comment (PowerfulBacon) [inline code/datums/chatmessage.dm]: this is pointless

File: code/datums/chatmessage.dm [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+		if(ishuman(target))
+			var/mob/living/carbon/human/H = target
+			if(H.wear_id?.GetID())
+				var/obj/item/card/id/idcard = H.wear_id
+				var/datum/job/wearer_job = SSjob.GetJob(idcard.GetJobName())
+				if(wearer_job)
+					target.chat_color = wearer_job.chat_color
+				else
+					target.chat_color = GLOB.job_colors_pastel[idcard.GetJobName()]
+				target.chat_color_name = H.name
```
Comment (PowerfulBacon) [inline code/datums/chatmessage.dm:137]: Is there even any point of setting this, since it isn't even used

Comment (Archanial) [reply code/datums/chatmessage.dm:137]: There is. If target stops being carbon/human it would be used.

File: code/datums/chatmessage.dm (hunk 15 above)
Comment (Archanial) [reply code/datums/chatmessage.dm]: my bad

File: code/datums/chatmessage.dm (hunk 17 above)
Comment (PowerfulBacon) [reply code/datums/chatmessage.dm:137]: If target stops being carbon/human then their variables are reset. If the variable aren't reset and the chat_color copies over then this entire system breaks.

**Comment by Crossedfall** [comment]
From the in-game poll (40 people voted no playtime restrictions were added)
image

**Comment by Vexylius** [comment]
Merge when?

**Comment by Archanial** [comment]
> 
> 
> Merge when?

When bacon reviews it and approves the changes, since he's code owner

**Review by francinum — APPROVED** [review]
has functioned just fine over several test merges and has community support.

**Review by PowerfulBacon — APPROVED** [review]
Shuttlecode changes seem fine

**Comment by Vexylius** [comment]
Mergeplz

**Review by utilbird — APPROVED** [review]
no

**Comment by francinum** [comment]
Not really this is good shit.

**Comment by jupyterkat** [comment]
@francinum duplicate define on the pr

**Comment by bobbah** [comment]
image

File: code/datums/emotes.dm [hunk 18]
```diff
@@ -73,13 +71,13 @@
 		if(!M.client || isnewplayer(M))
 			continue
```
Comment (bloons3) [inline code/datums/emotes.dm:72]: I'm gonna make fun of you for MONTHS about this

File: code/datums/emotes.dm [hunk 19]
```diff
@@ -61,9 +61,7 @@
 	if(!(end in list("!", ".", "?", ":", "\"", "-")))
 		msg += "."
 
-	user.log_message(msg, LOG_EMOTE)
```
Comment (bloons3) [inline code/datums/emotes.dm:64]: :)
